Find mallocMC: Prefer Paths over Default Paths

ax3l opened the pull request:
As documented in the `find_package` manual, adding `NO_DEFAULT_PATH`
allows to prefer PATHS over defaults such as the environment's
~~`MALLOCMC_ROOT` or the~~ `CMAKE_PREFIX_PATH` (both can be
set by users or a module system and might contain a valid,
but outdated version of mallocMC).
  http://www.cmake.org/cmake/help/v3.0/command/find_path.html

Between two `find_packages`, the user should clean the cached
result from it's `find_path` via

``` cmake
find_package(mallocMC 2.1.0 QUIET)

if(NOT mallocMC_FOUND)
  # remove cache results, e.g., of an outdated version
  unset(mallocMC_ROOT_DIR CACHE)

  message(STATUS "Using mallocMC from thirdParty/ directory")
  set(MALLOCMC_ROOT "${CMAKE_CURRENT_SOURCE_DIR}/../../thirdParty/mallocMC")
  find_package(mallocMC 2.1.0 REQUIRED)
endif(NOT mallocMC_FOUND)
```

see: http://stackoverflow.com/a/16309871

The CMake `MALLOCMC_ROOT` is still preferred over the environment `MALLOCMC_ROOT` (which is a good thing).

ax3l committed:
Find mallocMC: Prefer Hints over Default Paths

As documented in the `find_package` manual, adding `NO_DEFAULT_PATH`
allows to prefer PATHS over defaults such as the environment's
`MALLOCMC_ROOT` or the `CMAKE_PREFIX_PATH` (both can be
set by users or a module system and might contain a valid,
but outdated version of mallocMC).
  http://www.cmake.org/cmake/help/v3.0/command/find_path.html

Between two `find_packages`, the user should clean the cached
result from it's `find_path` via
```cmake
find_package(mallocMC 2.1.0 QUIET)

if(NOT mallocMC_FOUND)
  # remove cache results, e.g., of an outdated version
  unset(mallocMC_ROOT_DIR CACHE)

  message(STATUS "Using mallocMC from thirdParty/ directory")
  set(MALLOCMC_ROOT "${CMAKE_CURRENT_SOURCE_DIR}/../../thirdParty/mallocMC")
  find_package(mallocMC 2.1.0 REQUIRED)
endif(NOT mallocMC_FOUND)
```

see: http://stackoverflow.com/a/16309871

ax3l commented:
@slizzered can you pls take a look again? :)

ax3l commented:
note: I did not unset `mallocMC_ROOT_DIR` in the module itself (on `NOT mallocMC_FOUND`) since this messes up the caches and always _forces_ new `find_path` runs. (~~Not found is also a valid and cache-able result for multiple calls with the same information to the module.~~ -> `Once one of the calls succeeds the result variable will be set and stored in the cache so that no call will search again.`)

Since this var is exposed to the user, and the user also provides the new information before the next `find_package` call, I think user-level `unset` of the cache is the right way to do this.

ax3l committed:
Fix CACHE Scopes and Clean Ups

ax3l commented:
@slizzered pushed the updates. can you cross-check it with a valid halloc install pls? (compile, `mallocMC_halloc_FOUND`, `ccmake .` + `<t>`, etc.)

slizzered commented on FindmallocMC.cmake:
```diff
@@ -158,5 +167,31 @@ find_package_handle_standard_args(mallocMC
+
+    # user-level component vars
+    foreach(COMPONENT ${mallocMC_FIND_COMPONENTS})
+        # to do: expose to user in doc (header),
```
TODO is done and can be removed

slizzered commented on FindmallocMC.cmake:
```diff
@@ -158,5 +167,31 @@ find_package_handle_standard_args(mallocMC
+
+# clean up
+#
+# unset cached variables in case we did not found a valid install
```
typo: "found" should be "find"

slizzered commented on FindmallocMC.cmake:
```diff
@@ -97,44 +102,47 @@ if(mallocMC_ROOT_DIR)
             #
             list(APPEND mallocMC_REQUIRED_VARS_LIST mallocMC_LIBRARIES)
-            find_library(${COMPONENT}_LIBRARY
+            find_library(mallocMC_${COMPONENT}_LIBRARY
```
Should there also be a preliminary call to `find_library` that uses `NO_DEFAULT_PATH`?
And the second call of course without the `PATHS` option.

ax3l replied:
oh that is true!

slizzered commented on FindmallocMC.cmake:
```diff
@@ -97,44 +102,47 @@ if(mallocMC_ROOT_DIR)
             # halloc headers ########################################################
             #
-            find_path(${COMPONENT}_INCLUDE_DIR
+            find_path(mallocMC_${COMPONENT}_INCLUDE_DIR
```
Sould there also be one call to `find_path` that uses the `NO_DEFAULT_PATH` option?
And again, the second call does not need to use the `PATHS` variable anymore

ax3l replied:
yep

slizzered commented on FindmallocMC.cmake:
```diff
@@ -97,44 +102,47 @@ if(mallocMC_ROOT_DIR)
-            find_path(${COMPONENT}_INCLUDE_DIR
+            find_path(mallocMC_${COMPONENT}_INCLUDE_DIR
                 NAMES halloc.h
                 PATHS "${mallocMC_ROOT_DIR}/../halloc/" ENV HALLOC_ROOT
```
This line could be improved by  making it
`PATHS ${HALLOC_ROOT} "${mallocMC_ROOT_DIR}/../halloc/" ENV HALLOC_ROOT`
so there is also the CMake-Variable `HALLOC_ROOT` (simliar to `MALLOCMC_ROOT`)

ax3l replied:
and it also needs the two-fold `NO_DEFAULT_PATH` call. (update: ah you wrote it above)

ax3l replied:
hm, actually we would expect them to be installed in the same dir, so already the env is a bit weird (I am more tending to remove the `ENV HALLOC_ROOT` than adding just one more.

or do you think installing a mallocMC + halloc in two different locations makes it more useable (I doubt it from a useability perspective).

slizzered replied:
I think the real question is "do we want to redistribute _halloc_?".
Currently, the halloc code is not touched / split into policies. So there is no real reason why the whole source code should be shipped with mallocMC directly. Having halloc as an optional dependency would be more suitable. At least that would be the way to handle it if this was about 2 ArchLinux packages ;)

ax3l replied:
All right sounds very reasonable, then we go with that.

slizzered commented on FindmallocMC.cmake:
```diff
@@ -97,44 +102,47 @@ if(mallocMC_ROOT_DIR)
     #
     set(mallocMC_VERSION "${mallocMC_VERSION_MAJOR}.${mallocMC_VERSION_MINOR}.${mallocMC_VERSION_PATCH}")
-    set(mallocMC_INCLUDE_DIRS ${mallocMC_ROOT_DIR}/include)
+    set(mallocMC_INCLUDE_DIRS ${mallocMC_ROOT_DIR}/include CACHE PATH "mallocMC include pathes")
```
typo: pathes -> paths

slizzered replied:
How MPI handles the "problem" with cached variables:
they use a "work" variable inside their script and at the end, they create the cache variable like this:

``` cmake
 set(MPI_${lang}_COMPILE_FLAGS ${MPI_COMPILE_FLAGS_WORK} CACHE STRING "MPI ${lang} compilation flags" FORCE)
```

ax3l replied:
probably we can keep them non-cached, there is no direct use for the user to "modify" version and libraries/includes by hand (well, a bit for the latter) but since they are all printed that is good enough.

ax3l replied:
what variables caused the problem again? list of paths? (_LIBRARY _INCLUDE_DIR?)

slizzered replied:
Problems:
- mallocMC_LIBRARIES is not cached, whereas mallocMC_INCLUDE_DIRS is. This is inconsistent
- It is not possible to use `list(APPEND ...)` for mallocMC_INCLUDE_DIRS (list commands can not handle cached variables).

slizzered commented on FindmallocMC.cmake:
```diff
@@ -97,44 +102,47 @@ if(mallocMC_ROOT_DIR)
+                DOC "Libraries for the mallocMC component ${COMPONENT}."
+            )
+            if(mallocMC_${COMPONENT}_LIBRARY)
+                list(APPEND mallocMC_LIBRARIES ${mallocMC_${COMPONENT}_LIBRARY})
```
It seems that `list(APPEND)` can not directly create cached variables.
We need to use `set(mallocMC_LIBRARIES ${mallocMC_LIBRARIES} ${mallocMC_${COMPONENT}_LIBRARY} CACHE PATH "Libraries for the mallocMC compontent ${COMPONENT}")

Otherwise, the mallocMC_LIBRARIES variable is not visible within ccmake

slizzered commented on FindmallocMC.cmake:
```diff
@@ -97,44 +102,47 @@ if(mallocMC_ROOT_DIR)
+                DOC "Includes for the mallocMC component ${COMPONENT}."
+            )
+            if(mallocMC_${COMPONENT}_INCLUDE_DIR)
+                list(APPEND mallocMC_INCLUDE_DIRS ${mallocMC_${COMPONENT}_INCLUDE_DIR})
```
It looks like this call to `list(APPEND...)` does **not** do anything, presumably because it can not append to a cache variable?
`mallocMC_halloc_INCLUDE_DIR = /home/carli/dev/picongpu/thirdParty/mallocMC/halloc/src`
but in the end,
`mallocMC_INCLUDE_DIRS = /home/carli/dev/picongpu/thirdParty/mallocMC/src/include` (the new include_dir is missing).
Probably this has to be done using FORCE to update the cached variable?

slizzered commented on FindmallocMC.cmake:
```diff
@@ -158,5 +167,31 @@ find_package_handle_standard_args(mallocMC
+# (e.g., we only found an outdated version)
+if(NOT mallocMC_FOUND)
+    # default vars
+    unset(mallocMC_VERSION CACHE)
```
mallocMC_VERSION is not a cached variable

ax3l committed:
Fix Cache Weirdness (Last Commit)

Lists in caches cause trouble and versions
in the cache are not useful.

slizzered commented:
Not perfect, but probably OK for now.

One thing that I came across: All the cached variables that origin from calls to `find_path` and `find_library` are marked as _advanced_ per default.

ax3l commented:
cool, makes sense somehow.

slizzered commented on FindmallocMC.cmake:
```diff
@@ -158,5 +178,29 @@ find_package_handle_standard_args(mallocMC
+
+    # user-level component vars
+    foreach(COMPONENT ${mallocMC_FIND_COMPONENTS})
+        if(mallocMC_${COMPONENT}_FOUND)
```
This should happen unconditionally (not just when the component was found). Reason: something from the component (like the library) was found, but something else was not (like the include_dir). In this case, the code above will **unset** the `FOUND` variable.
As a result, the previously found part of the compontent (the library in this example) is not deleted!

ax3l replied:
do you mean, `the code above will unset the [component's] FOUND variable.` ? in that case, true! well spotted

slizzered replied:
yes, see [the code where it checks if the halloc include_dir was found](https://github.com/ComputationalRadiationPhysics/cmake-modules/blob/topic-mallocPreferHints/FindmallocMC.cmake#L152-L156) for an example of this actually happening.

ax3l replied:
marvelous, it seems that components make things easier, don't they? ;)

slizzered replied:
Yes, CMake did a wonderful job to ease the job of a developer that wants to use their component feature :poodle:

ax3l commented:
@slizzered something in this module (already the version currently checked in into [picongpu/dev](https://github.com/ComputationalRadiationPhysics/picongpu/blob/2d192167a63d06b3daf1b815bd01fc7c47a4b0bd/thirdParty/cmake-modules/FindmallocMC.cmake)) is not working with cmake `2.8.10.2`...

Do you remember what pushed the requirement?
(Can we avoid it? Else, I have to update the requirements in PIConGPU to that version, probably 3.x).

ax3l commented:
the options in `find_package_handle_standard_args`, especially components handling, was modified in [2.8.8](http://gccxml.org/Bug/bug_relationship_graph.php?bug_id=5920&graph=dependency&orientation=vertical)... that is earlier...

ok, `HANDLE_COMPONENTS` is interpreted as a var in the required_vars_list instead as the new option... when was it introduced?

slizzered commented:
2.8.8 introduced it, afaics
[2.8.7 does not have it](http://www.vtk.org/Wiki/CMake_2.8.7_Docs)
[2.8.8 uses `HANDLE_COMPONENTS`](http://www.vtk.org/Wiki/CMake_2.8.8_Docs)

For the reason why it doesn't work with 2.8.10, I don't know yet

slizzered commented:
something that was introduced in 2.8.11:
[FindCUDA: Add support for separable compilation](http://www.kitware.com/blog/home/post/492)

ax3l commented:
hm, but that does not cause the problem. the problem is, that in `2.8.10.2` the `_ROOT_DIR` is `...-NOTFOUND` but `mallocMC_FOUND` is true.

slizzered commented:
`_ROOT_DIR` is an element of the required vars list, so it should complain about it.

I think it could be related to `find_package_handle_standard_args`. The `FOUND_VAR` parameter is introduced in ~~CMake 3.0~~ 2.8.11

For previous versions, it looks like the name of the variable is fixed and you must omit the FOUND_VAR stuff

ax3l commented:
ah, that's what I am looking for!
actually introduced in [2.8.11](http://www.kitware.com/blog/home/post/492) (found no further fixes for it in [2.8.11.1](http://www.kitware.com/blog/home/post/507) and [2.8.11.2](http://www.kitware.com/blog/home/post/521))

ax3l committed:
CMake: FOUND_VAR & CUDA_SEPARABLE_COMPILATION

... require CMake 2.8.11

2.8.12   "FindCUDA: Multiple improvements to the custom commands"
2.8.12.1 "FindCUDA learned of CUDA 5.5's divided NPP libraries"
2.8.12.2 "FindOpenMP: Support compilers that do not need any special flags"

-> we use 2.8.12.2

ax3l commented:
huh,
- [2.8.12](http://www.kitware.com/blog/home/post/564) "FindCUDA: Multiple improvements to the custom commands" - whatever that means ;)
- [2.8.12.1](http://www.kitware.com/blog/home/post/571) "FindCUDA learned of CUDA 5.5's divided NPP libraries"
- [2.8.12.2](http://www.kitware.com/blog/home/post/620) _nothing_ relevant ... besides "FindOpenMP: Support compilers that do not need any special flags" ... hm, we make use of that now, too

ax3l commented:
@PrometheusPi thx for reporting the problem with cmake `2.8.10.2`!
